added native sensor_msgs::PointCloud2 support without conversion

spuetz opened the pull request:
Since the conversion from PointCloud2 to PointCloud and back for assembling the clouds is bad, I added a native sensor_msgs::PointCloud2 support without conversion. I also added an organized mode, to assemble PointClouds2 to an organized PointCloud2.

vrabaud commented on src/point_cloud2_assembler.cpp:
```diff
@@ -65,12 +76,10 @@ class PointCloud2Assembler : public BaseAssembler<sensor_msgs::PointCloud2>
-    sensor_msgs::convertPointCloud2ToPointCloud(scan_in, cloud_in);
-    tf_->transformPointCloud(fixed_frame_id, cloud_in, cloud_out) ;
-    return ;
+    pcl_ros::transformPointCloud(fixed_frame_id, cloud_in, cloud_out, *tf_);
```
pcl is a huge dependency for that simple function. Use tf2_sensor_msgs instead: https://github.com/ros/geometry_experimental/blob/indigo-devel/tf2_sensor_msgs/include/tf2_sensor_msgs/tf2_sensor_msgs.h#L59

vrabaud commented on include/laser_assembler/base_assembler.h:
```diff
@@ -88,7 +90,7 @@ class BaseAssembler
    * \param cloud_out The result of transforming scan_in into a cloud in frame fixed_frame_id
    */
-  virtual void ConvertToCloud(const std::string& fixed_frame_id, const T& scan_in, sensor_msgs::PointCloud& cloud_out) = 0 ;
+  virtual void Convert(const std::string& fixed_frame_id, const T& scan_in, V& cloud_out) = 0 ;
```
removing ConvertToCloud breaks the API: mark it as deprecated.

vrabaud commented on include/laser_assembler/base_assembler.h:
```diff
@@ -56,7 +58,7 @@ namespace laser_assembler
  * \brief Maintains a history of point clouds and generates an aggregate point cloud upon request
  */
-template<class T>
+template<class T, class V>
```
that breaks the API too unforttunately, you probably need to create another base class from which BaseAssembler would depend on

spuetz replied:
I don't understand in which way this should be done. Wouldn't it be to much overhead? Maybe a good way is to create a deprecated class: base_assembler with V=sensor_msgs::PointCloud What do you think?

vrabaud commented:
overall very useful, thank you. It will probably take use a few days to get through the code. I believe some code should be part of the base class (as private lightweight functions). E.g., all that code is duplicated for PointCloud and PointCloud2:

```
if (field->offset != this->scan_hist_[start_index].fields[chan_ind].offset){
```

Please have only one function for those (like a function that return analyses two different PointFields)

vrabaud commented:
@spuetz , any comments on that ?

spuetz commented:
@vrabaud, I added the method hasEqualScanHeader to check the headers. as recommended by you. Additionally, I fixed the down-sampling for the organized mode.

spuetz commented:
@vrabaud, would be nice, if we can complete this pull request and merge it. thx
